### PR TITLE
Implement GeoTileGrid aggregation

### PR DESCRIPTION
**Pull request by codebrain**

### CODE REVIEW

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+		internal override void WrapInContainer(AggregationContainer c) => c.GeoTile = this;
+	}
+
+	public class GeoTileGridAggregationDescriptor<T>
+		: BucketAggregationDescriptorBase<GeoTileGridAggregationDescriptor<T>, IGeoTileGridAggregation, T>
+			, IGeoTileGridAggregation
+		where T : class
+	{
+		/// <inheritdoc cref="IGeoTileGridAggregation.Field" />
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs]: `///inheritdoc` superfluous on explicit implemented properties (it already has this doc)

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		Field IGeoTileGridAggregation.Field { get; set; }
+
+		/// <inheritdoc cref="IGeoTileGridAggregation.Precision" />
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs]: `///inheritdoc` superfluous on explicit implemented properties (it already has this doc)

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		GeoTilePrecision? IGeoTileGridAggregation.Precision { get; set; }
+
+		/// <inheritdoc cref="IGeoTileGridAggregation.ShardSize" />
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs]: `///inheritdoc` superfluous on explicit implemented properties (it already has this doc)

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		int? IGeoTileGridAggregation.ShardSize { get; set; }
+
+		/// <inheritdoc cref="IGeoTileGridAggregation.Size" />
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTileGridAggregation.cs]: `///inheritdoc` superfluous on explicit implemented properties (it already has this doc)

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		Precision0 = 0,
+		Precision1 = 1,
+
+		/// <summary>
+		/// Subcontinental area
+		/// </summary>
+		Precision2 = 2,
+
+		/// <summary>
+		/// Largets Country
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: nitpick: Largest

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		/// </summary>
+		Precision13 = 13,
+		Precision14 = 14,
+
+		/// <summary>
+		/// Small Road
+		/// </summary>
+		Precision15 = 15,
+
+		// Street
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: `///<summary>Street</summary>`

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
+﻿namespace Nest
+{
+	public enum GeoTilePrecision
+	{
+		/// <summary>
+		/// Whole world
+		/// </summary>
+		Precision0 = 0,
+		Precision1 = 1,
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: needs a summary?

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+		/// <summary>
+		/// Subcontinental area
+		/// </summary>
+		Precision2 = 2,
+
+		/// <summary>
+		/// Largets Country
+		/// </summary>
+		Precision3 = 3,
+		Precision4 = 4,
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: needs a summary?

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		/// <summary>
+		/// Large European Country
+		/// </summary>
+		Precision6 = 6,
+
+		/// <summary>
+		/// Small Country / US State
+		/// </summary>
+		Precision7 = 7,
+		Precision8 = 8,
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: needs a summary

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+		/// <summary>
+		/// City / Town / District
+		/// </summary>
+		Precision12 = 12,
+
+		/// <summary>
+		/// Village / Suburb
+		/// </summary>
+		Precision13 = 13,
+		Precision14 = 14,
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: needs a summary?

File: src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+		Precision19 = 19,
+		Precision20 = 20,
+		Precision21 = 21,
+		Precision22 = 22,
+		Precision23 = 23,
+		Precision24 = 24,
+		Precision25 = 25,
+		Precision26 = 26,
+		Precision27 = 27,
+		Precision28 = 28,
```
Comment (russcam) [inline src/Nest/Aggregations/Bucket/GeoTileGrid/GeoTilePrecision.cs]: need summaries?

**Comment by russcam** [comment]
Can this be based against master, then ported to 7.1?

File: src/Tests/Tests/Aggregations/Bucket/GeoHashTile/GeoTileGridAggregationUsageTests.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+				Precision = GeoTilePrecision.Precision3,
+				Size = 1000,
+				ShardSize = 100
+			};
+
+		protected override void ExpectResponse(ISearchResponse<Project> response)
+		{
+			response.ShouldBeValid();
+			var myGeoTileTile = response.Aggregations.GeoTile("my_geotile");
+			myGeoTileTile.Should().NotBeNull();
```
Comment (Mpdreamz) [inline src/Tests/Tests/Aggregations/Bucket/GeoHashTile/GeoTileGridAggregationUsageTests.cs]: Can we assert some of the buckets on `myGeoTileTile` to be a ta more rigid?

**Comment by codebrain** [comment]
Aligning precision summaries with advice on this page: https://wiki.openstreetmap.org/wiki/Zoom_levels

File: src/Tests/Tests/Aggregations/Bucket/GeoHashTile/GeoTileGridAggregationUsageTests.cs (hunk 12 above)
Comment (codebrain) [reply src/Tests/Tests/Aggregations/Bucket/GeoHashTile/GeoTileGridAggregationUsageTests.cs]: Added some additional assertions.

**Pull request merged**